src/compiler/checker.ts

```diff
@@ -6899,10 +6899,15 @@ namespace ts {
         }
 
         function getConstraintOfIndexedAccess(type: IndexedAccessType) {
-            const objectType = getBaseConstraintOfType(type.objectType) || type.objectType;
-            const indexType = getBaseConstraintOfType(type.indexType) || type.indexType;
-            const constraint = !isGenericObjectType(objectType) && !isGenericIndexType(indexType) ? getIndexedAccessType(objectType, indexType, /*accessNode*/ undefined, errorType) : undefined;
-            return constraint && constraint !== errorType ? constraint : undefined;
+            const objectType = getConstraintOfType(type.objectType) || type.objectType;
+            if (objectType !== type.objectType) {
+                const constraint = getIndexedAccessType(objectType, type.indexType, /*accessNode*/ undefined, errorType);

```

Comment:
Wouldn't this still result in a type we may need to get the constraint of if the constraint of `tyoe.objectType` is a generic mapped type (since an index on a mapped type is trivially bounded by the mapped type's template)? Or is `getConstraintOfType` not the variant that can do that?

Reply:
If the _constraint_ of `type.objectType` is a mapped type, we do indeed fetch that here and form a new indexed access. If `type.objectType` _itself_ is a mapped type, the `getConstraintOfType` function does nothing, and we'll fall into the base constraint code below.